docs: Fixed Querier documentation for deduplication algorithm; Added FAQ (#5305)
thanos-io/thanos: docs/components/query.md

# Querier/Query

The `thanos query` command (also known as "Querier") primarily implements the server that exposes the [Prometheus HTTP v1 API](https://prometheus.io/docs/prometheus/latest/querying/api/) using Prometheus PromQL engine to query data in a Thanos cluster via PromQL.

In short, it gathers the data needed to evaluate the query from underlying [StoreAPIs](https://github.com/thanos-io/thanos/blob/main/pkg/store/storepb/rpc.proto), merges and deduplicate series, evaluates the query and returns the result. This allows "federated" global view of your metrics.

Apart from federating PromQL based Query APIs, Querier is capable of federating other metric specific endpoints:

* Labels
* Exemplars
* Scrape Targets
* Alerts and Rules
* Metadata

> > Querier is fully stateless and horizontally scalable. Querier is designed to work well no matter if you use sidecar, receiver or mixed Thanos deployment.

Reviewer: ```suggestion
> Querier is fully stateless and horizontally scalable. Querier is designed to work equally well with sidecar, receiver or mixed Thanos deployment.
```

Example command to run Querier:

```bash
thanos query \
    --http-address     "0.0.0.0:9090" \
    --store            "<store-api>:<grpc-port>" \
    --store            "<store-api2>:<grpc-port>" \
    --rule             "<rule-api>:<grpc-port>" \
    --exemplar         "<exemplar-api>:<grpc-port>"
```

## Querier Use Cases

> Why do I need this component?

> Thanos Querier allows to aggregate and optionally deduplicate multiple data backends under single endpoint.

Reviewer: ```suggestion
Thanos Querier allows to aggregate and optionally deduplicate multiple data backends under a single endpoint.
```

* Expose multiple [ExemplarAPI](https://github.com/thanos-io/thanos/blob/8fdd90c34318f453e0fda9b16ab2b201ed2ee4aa/pkg/exemplars/exemplarspb/rpc.proto#L25) gRPC backends as standard [Query Exemplars API](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-exemplars)
* Expose multiple [TargetAPI](https://github.com/thanos-io/thanos/blob/f5174e098b30413501a6c930769e57228c13aa68/pkg/targets/targetspb/rpc.proto#L26) gRPC backends as standard [Targets API](https://prometheus.io/docs/prometheus/latest/querying/api/#targets)
* Expose multiple [RulesAPI](https://github.com/thanos-io/thanos/blob/435700296491133dda9a2978bab46d0ba2e4e4f3/pkg/rules/rulespb/rpc.proto#L26) gRPC backends as standard [Rules API](https://prometheus.io/docs/prometheus/latest/querying/api/#rules)
* Expose multiple [MetadataAPI](https://github.com/thanos-io/thanos/blob/af4ee3a09f6b00d280f5b6913b18864a595dc96f/pkg/metadata/metadatapb/rpc.proto#L23) gRPC backends as standard [Metric Metadata API](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata)

> > Are you missing an API that you might find useful to use? Let us know on GitHub Issues!

Author: Good, point, will add in later commit or PR, since this is all green already

* Prometheus (see [Sidecar](sidecar.md))
* Object Storage (see [Store Gateway](store.md))
* Global alerting/recording rules evaluations (see [Stateful Ruler](rule.md))
* Metrics received from Prometheus remote write streams (see [Receiver](receive.md))
* Another Querier (you can stack Queriers on top of each other!)
* Non-Prometheus systems:
  * e.g [OpenTSDB](../integrations.md#opentsdb-as-storeapi) or [RemoteRead](https://github.com/G-Research/thanos-remote-read)

Thanks to that, you can run queries through API, Thanos UI, Grafana UI or via alerting or recording rules that aggregate metrics from mix of those sources.

Some examples:

* `sum(cpu_used{cluster=~"cluster-(eu1|eu2|eu3|us1|us2|us3)", job="service1"})` might give you sum of CPU used inside all listed clusters for service `service1`. This will work even if those clusters run multiple Prometheus servers each. Querier will know which data sources to query.

* In single cluster you shard Prometheus functionally or have different Prometheus instances for different tenants. You can spin up Querier to have access to all of the instances within single Query evaluation.

### Use Case 2: Run-time deduplication of HA groups

> Prometheus is stateful and by design does not allow replicating its database. This means that increasing high availability by running multiple Prometheus replicas is not very easy to use. Simple load balancing will not work as for example after some crash, a replica might be up but querying such replica will result in small gap during the period it was down. You have a second replica that maybe was up, but it could be down in other moment (e.g. rolling restart), so requests based load-balancing on top of those two or more instances won't be accurate.

Reviewer: > This means that increasing high availability by running multiple Prometheus replicas is not very easy to use

Not very easy to use in what respect? Perhaps we can reword this to something like  

```
 well.
Prometheus is stateful and by design does not allow replicating its database. Therefore increasing high availability by running multiple Prometheus replicas can introduce some challenges from a querying perspective.
```

Author: sgtm!

> To solve this Thanos Querier pulls the data from both replicas, and deduplicate those signals, filling the gaps if any, transparently to the Querier consumer. Read more about deduplication process in [Deduplication](#deduplication) section.

Reviewer: ```suggestion
To solve this, Thanos Querier pulls the data from both replicas, and deduplicates those signals whilst filling any gaps transparently to the Querier consumer. Read more about the deduplication process in [Deduplication](#deduplication) section.
```

## Metric Query Flow Overview

<img src="../img/querier.svg" class="img-fluid" alt="querier-steps"/>

Overall QueryAPI exposed by Thanos is guaranteed to be compatible with [Prometheus 2.x. API](https://prometheus.io/docs/prometheus/latest/querying/api/). The above diagram shows what Querier does for each Prometheus query request.

See [here](../service-discovery.md) on how to connect Querier with desired StoreAPIs (or other APIs in fact).

### Deduplication

The query layer can deduplicate series that were collected from high-availability pairs of data sources such as Prometheus or Receivers. A fixed single or multiple replica labels has to be configured at the start of Querier. Those special labels will be removed from each merged series. If, after label removal the series are duplicated, the deduplicate process occurs. Two or more series that are only distinguished by the given replica label, will be merged into a single time series.

> In practice the deduplication algorithm assumes that the same data was collected in the same or different times. The goal of this algorithm is to provide roughly the most consistent sample interval and only use the duplicated series as a fallback when main series is not following it's normal interval. This allows hiding gaps in collection of a single data source (e.g Prometheus) if another instance have data for it. It also works well to hide storage 1:1 replication done by Receivers. More on that later.

Reviewer: ```suggestion
In practice, the deduplication algorithm assumes that the same data was collected in the same or different times. The goal of this algorithm is to provide roughly the most consistent sample interval and only use the duplicated series as a fallback when main series is not following its normal interval. This allows hiding gaps in collection of a single data source (e.g Prometheus) if another instance has data for it. It also works well to hide storage 1:1 replication done by Receivers. More on that later.
```

* `up{job="prometheus",env="2",cluster="2",receive_replica="C",replica="A"} 1`

This logic can also be controlled via parameter on QueryAPI. More details in [API section](#deduplication-replica-labels).

#### Penalty Algorithm

Current algorithm Querier (and Compactor) uses for deduplication process can be called `penalty based`. The implementation is available [here](https://github.com/thanos-io/thanos/blob/d218e605c2dddab041551278bc1a5f632640f791/pkg/dedup/iter.go#L39). Generally it uses "the most stable" iterator (so source of series, e.g. Prometheus replica). The stability is detected by taking the first sample from both iterators and assigning penalty of 5s to the slower. Now if stable replica suddenly gives no sample, or sample with older timestamps plus penalty we switch to second one and give penalty for first one. This allows smoothly switching to stable replicas and still assuring the most consistent sample frequency possible.

In details:

Before deduplication does anything we remove exactly the same TSDB chunks from StoreAPI results [here](https://github.com/thanos-io/thanos/blob/de0e3848ff6085acf89a5f77e053c555a2cce550/pkg/query/iter.go#L76).

> Then if deduplication and replica labels were configured `dedup.NewSeriesSet` is invoked. It's algorithm can be outlined as follows:

Reviewer 2: ```suggestion
If deduplication and replica labels were configured then `dedup.NewSeriesSet` is invoked. It's algorithm can be outlined as follows:
```
